### DISCUSSION TITLE
Axis indicator: Is there a more Magnum-like way to do this?

### DISCUSSION THREAD

**Discussion by asmaloney** [mosra/magnum#645] [Q&A]
Goal: Add an axis indicator in the upper-right corner of my window.

So after a bunch of head-scratching & fumbling about I have something that seems to be working-ish:

<img width="582" alt="Screen Shot 2024-07-23 at 12 33 47 PM" src="https://github.com/user-attachments/assets/80f64283-6fb9-4968-b5f9-ca6ebbfba91d">

Here's what I did:

1) created a mesh using `Magnum::Primitives::axis3D` and `Magnum::MeshTools::generateLines`
2) created a `LineGL3D` shader to use it
3) in the shader's draw I get just the camera's rotation, then set the scale & position manually:
    ```cpp
   mShader->setWidth( 6.0f );
   mShader->setSmoothness( 1.0f );
   mShader->setViewportSize( Magnum::Vector2{ inCamera.viewport() } );

   const auto cCameraRotation = ( inCamera.projectionMatrix() * inCamera.cameraMatrix() ).rotationShear();

   Magnum::Matrix4 projection{};
   projection[0] = { cCameraRotation[0], 0.0f };
   projection[1] = { cCameraRotation[1], 0.0f };
   projection[2] = { cCameraRotation[2], 0.0f };

   projection = Magnum::Matrix4::translation( { 0.85f, 0.85f, 0.0f } ) *
                Magnum::Matrix4::scaling( { 0.1f, 0.1f, 0.1f } ) * projection;

   mShader->setTransformationProjectionMatrix( projection );

   mShader->draw( mesh );
    ```

This works, but still feels a little bit off. I wonder if it's because I still have shearing from the camera's rotation? Is there a way to eliminate that?

Is there a better approach to this or to fixing UI objects to the camera's viewport?

Thank you!

**Comment by mosra** [comment]
I suspect the projection matrix messes it up somehow, usually the projection contains also some sort of aspect ratio adjustment (i.e., scaling), and `rotationShear()` undoes that to have the axes unit length, causing potentially weird skewing.

I think this could work better?

```cpp
Matrix4 projection = 
    inCamera.projectionMatrix() *
    Matrix4::translation( { 0.85f, 0.85f, 0.0f } ) *
    Matrix4::scaling( { 0.1f, 0.1f, 0.1f } ) *
    Matrix4::from(inCamera.cameraMatrix().rotationShear(), {});
```

For fixing objects relative to camera, parent them to the camera object. The axis indicator is a bit special tho, because it's fixed to the viewport but also needs to rotate based on the actual camera rotation. I don't think there's a way to do that without manually extracting and applying the rotation.

**Comment by asmaloney** [comment]
> I think this could work better?

Unfortunately that made it disappear completely.

I think I need an orthographic projection of the object with just the rotation (and nothing else) from the camera. Does that sound right?

I've tried a bunch of different things using `gramSchmidtOrthonormalize` and `orthographicProjection`, but I'm really just floundering around as I'm not great at matrix math.

Do I need to create an orthographic camera and render with it instead? And somehow link its rotation to that of my main camera?

> For fixing objects relative to camera, parent them to the camera object.

Got it. Thank you.

**Comment by mosra** [comment]
> Unfortunately that made it disappear completely.

Haha. I'm not saying I'm great at matrix math either. Maybe dropping the translation part could make it appear in the center at least, and then you could fiddle with the X/Y offset to place it appropriately? That's what I do even 14 years into this project, flipping signs and the multiplication order until something appears :D 

> I think I need an orthographic projection

I _thought_ that wouldn't look right, so I didn't even attempt that. A way to achieve that would be dropping the `inCamera.projectionMatrix() *` part, as the default (identity) matrix is an orthographic projection on its own already. But you'll likely arrive at something very close to your original attempt (which was also orthographic), where the axes point almost correct, but not quite.

You don't need to create a separate orthographic camera for that, everything the scene graph camera is doing is calculating this one matrix, so if you create it directly, you have everything.

**Comment by asmaloney** [comment]
> That's what I do even 14 years into this project, flipping signs and the multiplication order until something appears

😆 Well that makes me feel _slightly_ better about not being able to solve what (I think) is a "simple problem".

I played around with many different things and can't get what I have in my head onto the screen, so I'm going to leave it for now and continue my experimentation with Magnum as a Qt3D replacement... I'll leave the question open in case someone wants to appear with a solution in the future 😄 

Thanks for your help!
